src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,127 @@
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        head = new Node<>(null, null, null);
+        tail = new Node<>(head, null, null);
+        head.next = tail;
+    }
+
+    private static class Node<T> {
+        private T element;
+        private Node<T> prev;
+        private Node<T> next;
+
+        private Node(Node<T> prev, T element, Node<T> next) {
+            this.element = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail.prev, value, tail);
+        tail.prev.next = newNode;
+        tail.prev = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> previousNode = findPreviousNode(index);
+        Node<T> newNode = new Node<>(previousNode, value, previousNode.next);
+        previousNode.next.prev = newNode;
+        previousNode.next = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> targetNode = getNodeByIndex(index);
+        T oldValue = targetNode.element;
+        targetNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> targetNode = getNodeByIndex(index);
+        targetNode.next.prev = targetNode.prev;
+        targetNode.prev.next = targetNode.next;
+        size--;
+        return targetNode.element;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            if (isEqual(get(i), object)) {
+                remove(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findPreviousNode(int index) {
+        checkIndexForAdd(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + " Size: " + size);
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i <= index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
```

Comment:
The only remark is the implementation of getNodeByIndex.
'''
Don't iterate your nodes from the head to the tail or vice versa.
Pay attention to the index you receive as an input, you can iterate only the first or second half of the list, depending on the index value. That will boost your lists' performance.
'''